File: app/src/main/java/com/lifedawn/bestweather/weathers/dataprocessing/util/SunRiseSetUtil.java
package com.lifedawn.bestweather.weathers.dataprocessing.util;

import com.luckycatlabs.sunrisesunset.SunriseSunsetCalculator;
import com.luckycatlabs.sunrisesunset.dto.Location;

import java.time.LocalTime;
import java.time.ZonedDateTime;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.TimeZone;
import java.util.concurrent.TimeUnit;

public class SunRiseSetUtil {
	private SunRiseSetUtil() {
	}

	public static Map<Integer, SunRiseSetObj> getDailySunRiseSetMap(ZonedDateTime begin, ZonedDateTime end, double latitude,
	                                                                double longitude) {
		ZonedDateTime criteria = ZonedDateTime.of(begin.toLocalDateTime(), begin.getZone());
		int beginDay = criteria.getDayOfYear();
		final int endDay = end.getDayOfYear();

		Map<Integer, SunRiseSetObj> map = new HashMap<>();

		SunriseSunsetCalculator calculator = new SunriseSunsetCalculator(new Location(latitude, longitude),
				TimeZone.getTimeZone(begin.getZone().getId()));

		Calendar calendar = Calendar.getInstance(TimeZone.getTimeZone(begin.getZone().getId()));

		do {
			calendar.setTimeInMillis(criteria.toInstant().toEpochMilli());

			map.put(beginDay, new SunRiseSetObj(calculator.getOfficialSunriseCalendarForDate(calendar), calculator.getOfficialSunsetCalendarForDate(calendar)));

			criteria = criteria.plusDays(1);
			beginDay = criteria.getDayOfYear();
		} while (beginDay <= endDay);

		return map;
	}

	public static boolean isNight(Calendar compDate, Calendar sunRiseDate, Calendar sunSetDate) {
		long compH = TimeUnit.MILLISECONDS.toHours(compDate.getTimeInMillis());
		long sunRiseH = TimeUnit.MILLISECONDS.toHours(sunRiseDate.getTimeInMillis());
		long sunSetH = TimeUnit.MILLISECONDS.toHours(sunSetDate.getTimeInMillis());
		if (compH > sunSetH || compH < sunRiseH) {
			return true;
		} else
			return false;
	}

	public static class SunRiseSetObj {
		final Calendar sunrise;
		final Calendar sunset;

		public SunRiseSetObj(Calendar sunrise, Calendar sunset) {
			this.sunrise = sunrise;
			this.sunset = sunset;
		}

		public Calendar getSunrise() {
			return sunrise;
		}

		public Calendar getSunset() {
			return sunset;
		}
	}
}

File: app/src/main/java/com/lifedawn/bestweather/data/remote/weather/dataprocessing/util/SunRiseSetUtil.kt
package com.lifedawn.bestweather.data.remote.weather.dataprocessing.util

import com.luckycatlabs.sunrisesunset.SunriseSunsetCalculator
import com.luckycatlabs.sunrisesunset.dto.Location
import java.time.ZoneId
import java.time.ZonedDateTime
import java.util.*
import java.util.concurrent.TimeUnit

object SunRiseSetUtil {
    fun getDailySunRiseSetMap(
        begin: ZonedDateTime, end: ZonedDateTime, latitude: Double,
        longitude: Double
    ): Map<Int, SunRiseSetObj> {
        val zoneId = begin.zone
        val utc0TimeZone = ZoneId.of(TimeZone.getTimeZone("UTC").id)
        val realTimeZone = TimeZone.getTimeZone(zoneId.id)
        var beginUtc0ZonedDateTime = ZonedDateTime.of(begin.toLocalDateTime(), zoneId)
        var beginRealZonedDateTime = ZonedDateTime.of(begin.toLocalDateTime(), zoneId)
        var endUtc0ZonedDateTime = ZonedDateTime.of(end.toLocalDateTime(), zoneId)
        beginUtc0ZonedDateTime = beginUtc0ZonedDateTime.withZoneSameLocal(utc0TimeZone)
        endUtc0ZonedDateTime = endUtc0ZonedDateTime.withZoneSameLocal(utc0TimeZone)
        var beginDay: Long
        val endDay = TimeUnit.MILLISECONDS.toDays(endUtc0ZonedDateTime.toInstant().toEpochMilli())
        val map: MutableMap<Int, SunRiseSetObj> = HashMap()
        val calculator = SunriseSunsetCalculator(Location(latitude, longitude), realTimeZone)
        val calendar = Calendar.getInstance(realTimeZone)
        do {
            calendar.timeInMillis = beginRealZonedDateTime.toInstant().toEpochMilli()
            map[beginRealZonedDateTime.dayOfYear] = SunRiseSetObj(
                calculator.getOfficialSunriseCalendarForDate(calendar),
                calculator.getOfficialSunsetCalendarForDate(calendar)
            )
            beginUtc0ZonedDateTime = beginUtc0ZonedDateTime.plusDays(1)
            beginRealZonedDateTime = beginRealZonedDateTime.plusDays(1)
            beginDay = TimeUnit.MILLISECONDS.toDays(beginUtc0ZonedDateTime.toInstant().toEpochMilli())
        } while (beginDay <= endDay)
        return map
    }

    fun isNight(compDate: Calendar, sunRiseDate: Calendar?, sunSetDate: Calendar?): Boolean {
        if (sunRiseDate == null || sunSetDate == null)
            return false

        val compH = TimeUnit.MILLISECONDS.toHours(compDate.timeInMillis)
        val sunRiseH = TimeUnit.MILLISECONDS.toHours(sunRiseDate.timeInMillis)
        val sunSetH = TimeUnit.MILLISECONDS.toHours(sunSetDate.timeInMillis)
        return compH > sunSetH || compH < sunRiseH
    }

    class SunRiseSetObj {
        var zonedDateTime: ZonedDateTime? = null
        val sunrise: Calendar
        val sunset: Calendar

        constructor(sunrise: Calendar, sunset: Calendar) {
            this.sunrise = sunrise
            this.sunset = sunset
        }

        constructor(zonedDateTime: ZonedDateTime?, sunrise: Calendar, sunset: Calendar) {
            this.zonedDateTime = zonedDateTime
            this.sunrise = sunrise
            this.sunset = sunset
        }
    }
}